### PR TITLE
Tag-transform: correct naming and transform direction

### PR DESCRIPTION
**Pull request by Gustl22**
the transform direction is named wrong.
- Tile source tags -> VTM tags (match): **transform**: _returns one value or null_
- VTM tags (match) -> tiles source tags: **retransform**: _returns list with at least one element or null_

Background is, that multiple tags can be transformed to one VTM tag (match), but not in the opposite direction.

Both directions are needed:
E.g. add following to `default.xml` so can try to transform colors, e.g. [here](https://overpass-turbo.eu/s/GsU).
```
<!--###### TAG TRANSFORM ######-->
    <tag-transform k="roof:colour" v="grey" k-match="roof:colour" v-match="#00ff00" />
    <tag-transform k="roof:colour" v="red" k-match="building:colour" v-match="#00ff00" />
    <tag-transform k="building:colour" v="white" k-match="building:colour" v-match="#ff0000" />
```
Can test it with `OverpassTest` as `Mapsforge maps` colors are already converted or use other tags there. Note: `setTheme` must be called before creation of `BuildingLayer` to take effect.

### CODE REVIEW

**Comment by devemux86** [comment]
So implementation was erroneous?

If it's about direction then the `retransform` term does not help much understand what is going on, more sounds like a double transform.

**Comment by Gustl22** [comment]
> So implementation was erroneous?

Not as far as we use only one transform per tag. But [naming](https://dict.leo.org/german-english/retransform) is wrong directed in my opinion.

**Comment by devemux86** [comment]
Also can explain why need now multiple results from a transform process when it fundamentally works with one-to-one theme declarations.

Note: we try to have something like [TagTransform](https://wiki.openstreetmap.org/wiki/Osmosis/TagTransform) working inside the themes.

**Comment by devemux86** [comment]
> But [naming](https://dict.leo.org/german-english/retransform) is wrong directed in my opinion.

Why need a reverse direction?

**Comment by Gustl22** [comment]
I all described in the first post?

**Comment by devemux86** [comment]
When have free time can test first current `master` implementation to see where are the errors.
Checking quickly the PR code cannot understand what is the purpose for existing tile sources.
Or all these are just for Overpass?

**Comment by Gustl22** [comment]
E.g. two color / material strings can end in the same color value.

**Comment by devemux86** [comment]
> E.g. two color / material strings can end in the same color value.

Is this about having color constants in themes or recognize from color values the materials?

**Comment by Gustl22** [comment]
> Or all these are just for Overpass?

No it's not meant for Overpass only. Generally when having two similar tags of tile source by design can end in the same VTM tag.

Currently we have no option to define the color strings ourselfs. We should then reverse it in Mapsforge maps creation, too, or use a tag transform there. But that's too widespread for now and only an example.

> Is this about having color constants in themes

Yes, at least for customizing hardcoded building colors.

But thats not the aim of this PR. Only an example.

**Comment by Gustl22** [comment]
I wanted to make this change before release, as it can cause compatibility issues.

**Comment by Gustl22** [comment]
> Why need a reverse direction?

Current implementation is already the reverse direction, but wrong named. So we need the _normal_ direction, too.

File: vtm/src/org/oscim/theme/XmlThemeBuilder.java [hunk 1]
```diff
@@ -1299,9 +1299,9 @@ private void tagTransform(String localName, Attributes attributes) {
         }
 
         if (v == null && matchValue == null) {
-            mTransformKeyMap.put(matchKey, k);
+            mTransformKeyMap.put(k, matchKey);
         } else {
-            mTransformTagMap.put(new Tag(matchKey, matchValue), new Tag(k, v));
+            mTransformTagMap.put(new Tag(k, v), new Tag(matchKey, matchValue));
```
Comment (devemux86) [inline vtm/src/org/oscim/theme/XmlThemeBuilder.java]: Ah you reversed the storage, that's why needed to reverse other things as well.

Comment (devemux86) [inline vtm/src/org/oscim/theme/XmlThemeBuilder.java]: Oh you reversed the storage, that's why then need to reverse all implementation as well.

Though cannot understand the reason for this change..
Normally need to find the match from the key, that's how is declared in themes and expected to be used.

**Comment by devemux86** [comment]
_Please avoid multiple PRs modifying the same classes, like with current and #675 on `BuildingLayer`. Becomes error prone and time consuming to review..._

File: vtm/src/org/oscim/theme/XmlThemeBuilder.java (hunk 1 above)
Comment (devemux86) [inline vtm/src/org/oscim/theme/XmlThemeBuilder.java]: Why the reverse (please explain the changes in detail) ?

**Comment by devemux86** [comment]
Sorry description remains vague, so cannot understand the concept of changes or need of them.

Also don't like the creation of  `List` for each tag, such often object creations should be avoided.
And what is that `res.get(0)`, all this trouble and then always get the 1st result in a created list?

The concept of introduce [TagTransform](https://wiki.openstreetmap.org/wiki/Osmosis/TagTransform) in VTM was because of providers modifying the OSM tags, so can (probably) use common rules / tags also with them, like with [mapzen](https://github.com/mapsforge/vtm/blob/master/vtm-themes/resources/assets/vtm/mapzen.xml#L6-L15) or [openmaptiles](https://github.com/mapsforge/vtm/blob/master/vtm-themes/resources/assets/vtm/openmaptiles.xml#L6-L9) themes.

Must first understand in detail what is wrong with our current concept, fix it if needed and afterwards what case cannot (if should) be supported, step by step.

**Comment by Gustl22** [comment]
> Also don't like the creation of List for each tag

Alternatively can assume, that first tag is always given in an element. So can avoid list creation.

> Must first understand in detail what is wrong with our current concept

Simply try to transform a Tile source tag to a VTM tag (inklusive value). It will not work in master.

**Comment by devemux86** [comment]
> Simply try to transform a Tile source tag to a VTM tag (inklusive value). It will not work in master.

So is there a problem in aforementioned mapzen or openmaptiles themes?

**Comment by Gustl22** [comment]
No. There was no case yet where needed to transform values as well.

**Comment by Gustl22** [comment]
At least we have to rename the `transform` to `retransform` or another term, as current implementation is the backward direction and not should conflict / mislead with possible implementation of forward transform.

File: vtm/src/org/oscim/theme/XmlThemeBuilder.java (hunk 1 above)
Comment (Gustl22) [reply vtm/src/org/oscim/theme/XmlThemeBuilder.java]: As said, rule is:
`k` is transformed to `matchKey`
`v` is transformed to `matchValue`

One `k` can only have one `matchKey`, as they're all processed in the same rule.
But one  `matchKey` can have multiple `k`.
See example in first post with colors, to understand. So the map is wrong directed to represent this relation.

Same for Tags with values.

**Comment by devemux86** [comment]
The idea is to allow tag transform at render themes. That happens in one direction inside the transform rules, source > target, where source is unique, is used in search and should be instant. If current implementation was done wrong obviously can improve / fix it.

**Comment by Gustl22** [comment]
> used in search and should be instant.

We already discussed that in #587 to not use the [instant/replace](https://github.com/mapsforge/vtm/pull/587#discussion_r216903249) option. So current is a lighter implementation which asks only for transform tags, when needed, so it's reversed approach. But for query values of reversed key, we need the forward transform option, too.

> If current implementation was done wrong obviously can improve / fix it.

So what is your concern on the PR or the further steps? Of course can take time to review.

**Comment by devemux86** [comment]
There seems to be an error in storage in `XmlThemeBuilder` and need to reverse the keys / values?

Anything extra where is needed and for what workflow that cannot be served with current implementation?

**Comment by Gustl22** [comment]
I don't understand your concern?

> There seems to be an error in storage in `XmlThemeBuilder` and need to reverse the keys / values?

Is it a hint or a question? As you can see yourself the map is initialized in the opposite direction of regular transform `matchKey` -> `key` means REVERSED directed. Or alternatively **naming** of methods is WRONG.

> cannot be served with current implementation

Having a s3db tile server like the one of opensciencemap, or Mapsforge maps with color strings, and want to transform them to customized color hex. Like shown in FIRST post.

So my concern before the release is to correct the naming cause of compatibility. The implementation can be reversed when ever.

**Comment by devemux86** [comment]
My concern is first that this PR seems to handle 2 things, possible wrong storage and second a new feature. These belong to separate PRs.

Also what do you mean to go from values to keys, so need lists? Why is that strange approach?

**Comment by devemux86** [comment]
> Is it a hint or a question? As you can see yourself the map is initialized in the opposite direction of regular transform `matchKey` -> `key` means REVERSED directed. Or alternatively **naming** of methods is WRONG.

You made the change and ask me?
Again if there is an error in `master` please post a separate simple isolated PR to fix it, after discussing what needs to be changed.
And naming should be simple like used already in [community](https://wiki.openstreetmap.org/wiki/Osmosis/TagTransform), as everything should be clearly understandable for the users.

**Comment by Gustl22** [comment]
> My concern is first that this PR seems to handle 2 things, possible wrong storage and second a new feature. These belong to separate PRs.

Agreed.

**Comment by devemux86** [comment]
For reference, like in #678 also here, a quick testing reveals that buildings rendering is broken:

**Pull request closed without merging**